File: src/pasta/toprf.rs
//! 2HashTDH — Jarecki et al. [JKKX17] の閾値 OPRF。PASTA が使っているもの。
//!
//! サーバ側の秘密 k を n 台に分散し、クライアントはパスワードを一切明かさずに
//! `h = F_k(password)` を得る。t 台の協力が無ければ誰も F_k を計算できないので、
//! 1 台侵害されただけではオフライン辞書攻撃が成立しない。
//!
//! PASTA の要求どおり **TOPRF 鍵はクライアントごとに生成する**。
//! 全クライアントで鍵を共有すると、1 人分の総当たりで集めた PRF 値を
//! 他のクライアントへのオフライン攻撃に流用されてしまう。

use curve25519_dalek::{RistrettoPoint, Scalar};
use sha2::{Digest, Sha512};

use super::shamir::{ParticipantId, Share, lagrange_coeff, random_scalar};

/// H2: X → G。パスワードを群の元へ写す。
fn hash_to_group(input: &[u8]) -> RistrettoPoint {
    let mut hasher = Sha512::new();
    hasher.update(b"PASTA-TOPRF-H2");
    hasher.update((input.len() as u64).to_le_bytes());
    hasher.update(input);
    let digest = hasher.finalize();

    let mut wide = [0u8; 64];
    wide.copy_from_slice(&digest);
    RistrettoPoint::from_uniform_bytes(&wide)
}

/// クライアント側の目隠し状態。`r` は復元時まで秘匿する。
pub struct Blinding {
    r: Scalar,
}

/// パスワードを目隠しして送信用の点を作る。a = [r]·H2(pw)
pub fn blind(password: &[u8]) -> (Blinding, RistrettoPoint) {
    let r = random_scalar();
    (Blinding { r }, hash_to_group(password) * r)
}

/// サーバ i の部分評価。b_i = [k_i]·a
///
/// サーバは `a` から password を復元できない（離散対数と目隠しのため）。
pub fn evaluate(key_share: &Share, blinded: &RistrettoPoint) -> RistrettoPoint {
    blinded * key_share.value
}

/// 部分評価をラグランジュ補間で結合し、目隠しを外す。z = H2(pw)^k
pub fn unblind(
    blinding: &Blinding,
    partials: &[(ParticipantId, RistrettoPoint)],
) -> RistrettoPoint {
    let ids: Vec<ParticipantId> = partials.iter().map(|(id, _)| *id).collect();

    let combined = partials
        .iter()
        .map(|(id, point)| point * lagrange_coeff(&ids, *id))
        .fold(RistrettoPoint::default(), |acc, p| acc + p);

    combined * blinding.r.invert()
}

/// H1: X × G → {0,1}^256。最終的な PRF 出力 h を得る。
pub fn finalize(password: &[u8], z: &RistrettoPoint) -> [u8; 32] {
    let mut hasher = Sha512::new();
    hasher.update(b"PASTA-TOPRF-H1");
    hasher.update((password.len() as u64).to_le_bytes());
    hasher.update(password);
    hasher.update(z.compress().as_bytes());
    let digest = hasher.finalize();

    let mut out = [0u8; 32];
    out.copy_from_slice(&digest[..32]);
    out
}

/// サーバ i に渡す派生値 h_i = H'(h, i)。
///
/// PASTA の client impersonation 対策。h そのものはどのサーバにも渡らないので、
/// 1 台侵害しても攻撃者が得るのはその台の h_i だけで、残りは秘密のまま。
pub fn derive_server_key(h: &[u8; 32], id: ParticipantId) -> [u8; 32] {
    let mut hasher = Sha512::new();
    hasher.update(b"PASTA-TOPRF-H-PRIME");
    hasher.update(h);
    hasher.update(id.to_le_bytes());
    let digest = hasher.finalize();

    let mut out = [0u8; 32];
    out.copy_from_slice(&digest[..32]);
    out
}

/// クライアント固有の TOPRF 鍵を作り、(t, n) で分散する。登録時に一度だけ実行。
pub fn generate_key(n: ParticipantId, t: ParticipantId) -> Vec<Share> {
    super::shamir::split(&random_scalar(), n, t)
}

#[cfg(test)]
mod tests {
    use super::*;

    /// どの t 台の組み合わせでも同じ h が得られる（＝決定的な PRF である）。
    #[test]
    fn any_quorum_yields_same_output() {
        let key_shares = generate_key(5, 3);
        let password = b"correct horse battery staple";

        let evaluate_with = |subset: &[usize]| {
            let (blinding, blinded) = blind(password);
            let partials: Vec<_> = subset
                .iter()
                .map(|&i| (key_shares[i].id, evaluate(&key_shares[i], &blinded)))
                .collect();
            finalize(password, &unblind(&blinding, &partials))
        };

        assert_eq!(evaluate_with(&[0, 1, 2]), evaluate_with(&[2, 3, 4]));
    }

    /// パスワードが違えば出力も違う。
    #[test]
    fn different_password_yields_different_output() {
        let key_shares = generate_key(3, 3);

        let run = |password: &[u8]| {
            let (blinding, blinded) = blind(password);
            let partials: Vec<_> = key_shares
                .iter()
                .map(|s| (s.id, evaluate(s, &blinded)))
                .collect();
            finalize(password, &unblind(&blinding, &partials))
        };

        assert_ne!(run(b"hunter2"), run(b"hunter3"));
    }

    /// t 未満では正しい h に到達できない。
    #[test]
    fn below_threshold_fails() {
        let key_shares = generate_key(5, 3);
        let password = b"hunter2";

        let (blinding, blinded) = blind(password);
        let full: Vec<_> = key_shares[0..3]
            .iter()
            .map(|s| (s.id, evaluate(s, &blinded)))
            .collect();
        let short: Vec<_> = key_shares[0..2]
            .iter()
            .map(|s| (s.id, evaluate(s, &blinded)))
            .collect();

        assert_ne!(
            finalize(password, &unblind(&blinding, &full)),
            finalize(password, &unblind(&blinding, &short))
        );
    }
}

File: projects/dealer/src/crypto/toprf.ts
import { ristretto255 } from '@noble/curves/ed25519';
import { sha512 } from '@noble/hashes/sha512';
import { hkdf } from '@noble/hashes/hkdf';
import { sha256 } from '@noble/hashes/sha256';
import {
  ED25519_ORDER,
  ParticipantId,
  Share,
  invert,
  lagrangeCoeff,
  mod,
  randomScalar,
  splitSecret,
} from './shamir.js';

export type RistrettoPoint = typeof ristretto255.Point.BASE;

export interface Blinding {
  r: bigint;
}

export interface PartialEvaluation {
  id: ParticipantId;
  point: RistrettoPoint;
}

function toBytes(input: Uint8Array | string): Uint8Array {
  return typeof input === 'string' ? new TextEncoder().encode(input) : input;
}

function u64ToBytesLE(n: number | bigint): Uint8Array {
  const buf = new Uint8Array(8);
  const view = new DataView(buf.buffer);
  view.setBigUint64(0, BigInt(n), true);
  return buf;
}

function u16ToBytesLE(n: number): Uint8Array {
  const buf = new Uint8Array(2);
  const view = new DataView(buf.buffer);
  view.setUint16(0, n, true);
  return buf;
}

/**
 * H_1: Maps a password to a point on Ristretto255 curve.
 * Hash to uniform 64 bytes with SHA-512, then map to Ristretto255 group.
 */
export function hashToGroup(password: Uint8Array | string): RistrettoPoint {
  const pw = toBytes(password);
  const prefix = new TextEncoder().encode('PASTA-TOPRF-H2');
  const lenBytes = u64ToBytesLE(pw.length);

  const input = new Uint8Array(prefix.length + lenBytes.length + pw.length);
  input.set(prefix, 0);
  input.set(lenBytes, prefix.length);
  input.set(pw, prefix.length + lenBytes.length);

  const digest = sha512(input);
  return ristretto255.Point.hashToCurve(digest);
}

/**
 * Client-side blinding:
 * Computes A = r * H_1(pw) where r is a secret random scalar.
 *
 * @param password User's password
 * @returns Secret blinding factor `r` and blinded group point `A`.
 */
export function blind(password: Uint8Array | string): {
  blinding: Blinding;
  blinded: RistrettoPoint;
} {
  const r = randomScalar();
  const h1 = hashToGroup(password);
  const A = h1.multiply(r);
  return { blinding: { r }, blinded: A };
}

/**
 * Server i partial evaluation:
 * Computes B_i = k_i * A
 * The server learns nothing about the password due to the blinding factor r and discrete log.
 *
 * @param keyShare Server i's share of the TOPRF key
 * @param blinded The blinded point A from client
 */
export function evaluate(keyShare: Share, blinded: RistrettoPoint): RistrettoPoint {
  return blinded.multiply(keyShare.value);
}

/**
 * Client-side unblinding:
 * Combines partial evaluations using Lagrange interpolation:
 * C = ∑ λ_i * B_i
 * v = r^(-1) * C = H_1(pw)^k
 *
 * @param blinding Blinding factor containing secret r
 * @param partials Array of partial evaluations from servers (id and B_i)
 */
export function unblind(
  blinding: Blinding,
  partials: Array<{ id: ParticipantId; point: RistrettoPoint }>
): RistrettoPoint {
  if (partials.length === 0) {
    throw new Error('At least one partial evaluation is required');
  }

  const ids = partials.map((p) => p.id);
  let combined = ristretto255.Point.ZERO;

  for (const partial of partials) {
    const lambda = lagrangeCoeff(ids, partial.id);
    const term = partial.point.multiply(lambda);
    combined = combined.add(term);
  }

  const rInv = invert(blinding.r, ED25519_ORDER);
  return combined.multiply(rInv);
}

/**
 * H_2: X × G -> {0, 1}^256.
 * Finalizes PRF value: h = H_2(password, v).
 *
 * @param password User's password
 * @param v Unblinded group point
 * @returns 32-byte master PRF output h
 */
export function finalize(password: Uint8Array | string, v: RistrettoPoint): Uint8Array {
  const pw = toBytes(password);
  const prefix = new TextEncoder().encode('PASTA-TOPRF-H1');
  const lenBytes = u64ToBytesLE(pw.length);
  const vBytes = v.toRawBytes();

  const input = new Uint8Array(prefix.length + lenBytes.length + pw.length + vBytes.length);
  let offset = 0;
  input.set(prefix, offset);
  offset += prefix.length;
  input.set(lenBytes, offset);
  offset += lenBytes.length;
  input.set(pw, offset);
  offset += pw.length;
  input.set(vBytes, offset);

  const digest = sha512(input);
  return digest.slice(0, 32);
}

/**
 * H_3: Derives server i-specific key h_i = H_3(h, i).
 * Uses SHA-512 matching PASTA specification to prevent client impersonation.
 *
 * @param h Master PRF output (32 bytes)
 * @param id Server participant ID
 * @returns 32-byte server key h_i
 */
export function deriveServerKey(h: Uint8Array, id: ParticipantId): Uint8Array {
  const prefix = new TextEncoder().encode('PASTA-TOPRF-H-PRIME');
  const idBytes = u16ToBytesLE(id);

  const input = new Uint8Array(prefix.length + h.length + idBytes.length);
  input.set(prefix, 0);
  input.set(h, prefix.length);
  input.set(idBytes, prefix.length + h.length);

  const digest = sha512(input);
  return digest.slice(0, 32);
}

/**
 * Optional HKDF-based server key derivation (H_3 via HKDF-SHA256).
 */
export function deriveServerKeyHkdf(h: Uint8Array, id: ParticipantId): Uint8Array {
  const info = new Uint8Array(new TextEncoder().encode(`PASTA-TOPRF-SERVER-KEY-${id}`));
  return hkdf(sha256, h, undefined, info, 32);
}

/**
 * Generates a client-specific TOPRF key and shares it via (t, n) Shamir secret sharing.
 * Generated once per client during registration.
 *
 * @param total Number of servers (n)
 * @param threshold Threshold number of servers (t)
 */
export function generateToprfKey(total: number, threshold: number): Share[] {
  const masterKey = randomScalar();
  return splitSecret(masterKey, threshold, total);
}
